lib/cluster.js

```diff
@@ -198,10 +210,32 @@ function addListeners() {
 			if (message.type === GET_METRICS_REQ) {
 				Promise.all(registries.map(r => r.getMetricsAsJSON()))
 					.then(metrics => {
-						process.send({
-							type: GET_METRICS_RES,
-							requestId: message.requestId,
-							metrics,
+						metrics.forEach(registry => {
+							registry.forEach(value => {
+								const hash = hashObject(value);
+								const key = `${value.metricName}_${hash}`;
+								value.hash = key;
+							});
+						});
+						// adding request id in file path to handle concurrency
+						const filename = path.join(
+							os.tmpdir(),
+							`metrics-${process.pid}-${message.requestId}.json`,
+						);
+						fs.writeFile(filename, JSON.stringify(metrics), err => {
```

Comment:
I'm really hesitant to involve files, since that suddenly involves the page cache, file system, more syscalls and possibly other surprises like filesystem permissions, running out of disk space and cleaning up. How much of an improvement did this yield? Did you benchmark before or after Node.js v18.6.0, which has [this optimization](https://github.com/nodejs/node/pull/42931)? How much disk I/O load did your system have when benchmarking? I think IPC is "supposed" to be faster than involving the filesystem.

Reply:
The benchmarks that I have performed are on our own application. The screenshots of the results are shared in the
this [Issue Link](https://github.com/siimon/prom-client/issues/628). The benchmarks were done on a 16 core machine and approximately 250-300 metrics (total size ~500 Kbs). At P99 and above levels we are getting ~10x improvement in app performance. The major choking point was with IPC. The detailed bifurcation that I have done on my local machine which is 8 core machine is given In the screenshot below. and the code used is in below given zip file. Have added some consoles in existing promc-client in this zip. Here if we see the worst IPC time is always higher. I had run multiple iterations and same.similar results are found. haven't tested with node v18.6.0 yet. will test the same code with that as well and share the results in the same thread
[cluster_test.zip](https://github.com/user-attachments/files/15531600/cluster_test.zip)
<img width="348" alt="Screenshot 2024-06-03 at 2 27 48 PM" src="https://github.com/siimon/prom-client/assets/22372359/6e86e2de-d941-4249-8e0e-a20756ed70ed">

Reply:
In these benchmarks ([issue link](https://github.com/siimon/prom-client/issues/628)) the scraping interval used was every 5 seconds and the throughput on the app was about 1100-1200 RPS.

Reply:
Hi @zbjornson 

- I also tried the code in the zip file with node v18.6.0. I am getting the similar results without any improvements in the logged timings.
- I think we shouldn't use IPC for this communication because that can create a problem whenever the sizes of metrics go beyond a certain point and that will start blocking the requests routing at high throughputs. 
- We can ask user to provide permissions for /tmp folder. From prom-client performance POV i didn't see much of a difference with files but again I am only hitting it once every 5 seconds.
- If not files then we should think of other ways of communication to solve this completely

Reply:
Hi @zbjornson @SimenB we went live in production with these changes about 3 months ago and everything is working fine for us. We are not seeing any additional latencies post this change. We are service at more than 100,000 RPS at peak for this service. If required I can share the production results as well.
I understand that introducing file system can pose different challenges for the library, But we can explore communication with HTTP or some other mode of communication (Not IPC) when we have larger size of the metrics as the current solution will cause very high tail latencies (>P99) at high throughput and larger metrics sizes. 
Requesting your inputs on this, In terms of how we can take this forward as we don't want to diverge from the library maintaining our own custom version.